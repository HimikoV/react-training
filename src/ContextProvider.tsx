import { createContext, useEffect, useState, useMemo } from "react";
import { mapBasketData } from "./mappers/mapBasketData";
import { getDeliveryOptions } from "./service/getDeliveryOptions";
import { getOrderData } from "./service/getOrderData";
import { getPaymentOptions } from "./service/getPaymentOptions";
import { requestData } from "./service/requestData";
import { basketType } from "./types/basketType";
import { finalizeOptionsType } from "./types/finalizeOptionsType";
import { orderType } from "./types/orderType";
import { productType } from "./types/productType";
import { childrenInterface } from "./interfaces/childrenInterface";
import { basketProcessedType } from "./types/basketProcessedType";
import { contextProviderInterface } from "./interfaces/contextProviderInterface";

export const dataContext = createContext<contextProviderInterface>(
  {} as contextProviderInterface
);

const ContextProvider = ({ children }: childrenInterface) => {
  const [basketData, setBasketData] = useState<basketType[]>(
    JSON.parse(localStorage.getItem("basketData") || "[]")
  );
  const [filter, setFilter] = useState<string[]>([]);
  const [data, setData] = useState<productType[]>([]);
  const [loadLimit, setLoadLimit] = useState<number>(9);
  const [basketProcessedData, setBasketProcessedData] = useState<
    basketProcessedType[]
  >([]);
  const [deliveryOptions, setDeliveryOptions] = useState<finalizeOptionsType[]>(
    []
  );
  const [paymentOptions, setPaymentOptions] = useState<finalizeOptionsType[]>(
    []
  );
  const [open, setOpen] = useState<boolean>(false);
  const [orderData, setOrderData] = useState<orderType[]>([]);

  useEffect(() => {
    const fetchData = async () => {
      const responseData = await requestData();
      setData(responseData);
      const responseDelivery = await getDeliveryOptions();
      setDeliveryOptions(responseDelivery);
      const responsePayment = await getPaymentOptions();
      setPaymentOptions(responsePayment);
      const responseOrderData = await getOrderData();
      setOrderData(responseOrderData);
    };
    fetchData();
  }, []);

  useEffect(() => {
    localStorage.setItem("basketData", JSON.stringify(basketData));
  }, [basketData]);

  useEffect(() => {
    mapBasketData(data, basketData, setBasketProcessedData);
  }, [data]);

  const trueValues = useMemo(
    () => ({
      filter,
      setFilter,
      basketData,
      setBasketData,
      data,
      loadLimit,
      setLoadLimit,
      basketProcessedData,
      setBasketProcessedData,
      deliveryOptions,
      setDeliveryOptions,
      paymentOptions,
      setPaymentOptions,
      open,
      setOpen,
      orderData,
      setOrderData,
    }),
    [
      filter,
      setFilter,
      basketData,
      setBasketData,
      data,
      loadLimit,
      setLoadLimit,
      basketProcessedData,
      setBasketProcessedData,
      deliveryOptions,
      setDeliveryOptions,
      paymentOptions,
      setPaymentOptions,
      open,
      setOpen,
      orderData,
      setOrderData,
    ]
  );

  // const values = {};
  return (
    <dataContext.Provider value={trueValues}>{children}</dataContext.Provider>
  );
};
export default ContextProvider;
