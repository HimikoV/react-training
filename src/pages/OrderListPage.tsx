import { Grid, Button, Typography } from "@mui/material";
import OrderList from "../components/orders/OrderList";
import OrderImageBg from "../components/orders/OrderImageBg";
import { dataContext } from "../ContextProvider";
import { useContext } from "react";

const OrderListPage = () => {
  const { orderLoadLimit, setOrderLoadLimit, orderData } =
    useContext(dataContext);
  return (
    <Grid
      sx={{
        display: "flex",
        flexWrap: "wrap",
        justifyContent: "flext-start",
        alignItems: "center",
        flexDirection: "column",
        position: "relative",
        minHeight: "100vh",
        color: "#3d405b",
        minWidth: "100%",
      }}
    >
      <OrderImageBg />
      <Typography variant="h3" color="primary">
        --Order List--
      </Typography>
      <OrderList />
      {orderData.length - orderLoadLimit > 0 && (
        <Button
          onClick={() => {
            setOrderLoadLimit(orderLoadLimit + 20);
          }}
          sx={{ my: 2 }}
        >
          <Typography variant="button" color="primary">
            Load More...
          </Typography>
        </Button>
      )}
    </Grid>
  );
};

export default OrderListPage;
