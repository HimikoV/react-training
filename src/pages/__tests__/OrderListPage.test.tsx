import OrderListPage from "../OrderListPage";
import { describe, it } from "@jest/globals";
import renderer from "react-test-renderer";
import ContextProvider from "../../ContextProvider";

describe("testing OrderListPage render", () => {
  it("OrderListPage Renders Correctly", () => {
    const orderListPage = renderer
      .create(
        <ContextProvider>
          <OrderListPage />
        </ContextProvider>
      )
      .toJSON();
    expect(orderListPage).toMatchSnapshot();
  });
});
