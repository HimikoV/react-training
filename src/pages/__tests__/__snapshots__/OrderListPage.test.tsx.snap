// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`testing OrderListPage render OrderListPage Renders Correctly 1`] = `
<div
  className="MuiGrid-root css-1xsjdn8-MuiGrid-root"
>
  <div
    className="MuiGrid-root css-b7ixwv-MuiGrid-root"
  >
    <img
      alt="BG for products Page"
      className="MuiBox-root css-oo92pk"
      src="https://images.unsplash.com/photo-1603984719659-7af2c6f02a79?ixlib=rb-1.2.1&ixid=MnwxMjA3fDB8MHxwaG90by1wYWdlfHx8fGVufDB8fHx8&auto=format&fit=crop&w=687&q=80"
    />
  </div>
  <h3
    className="MuiTypography-root MuiTypography-h3 css-1biy274-MuiTypography-root"
  >
    --Order List--
  </h3>
  <span
    className="MuiTypography-root MuiTypography-caption css-1d9s20a-MuiTypography-root"
  >
    There is no orders yet, please do a first order to see the history of orders
  </span>
</div>
`;
